xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,165 @@
+      final String authority) {
+    this.channel = checkNotNull(channel, "channel");
+    this.xdsNode = checkNotNull(xdsNode, "xdsNode");
+    this.authority = checkNotNull(authority, "authority");
+    this.endpointWatchers = new EndpointWatchers(xdsNode);
+    this.adsLifecycleManager = new AdsLifecycleManager(
+        AggregatedDiscoveryServiceGrpc.newStub(channel).withWaitForReady(),
+        endpointWatchers,
+        syncCtx,
+        checkNotNull(timerService, "timerService"),
+        checkNotNull(stopwatchSupplier, "stopwatchSupplier"),
+        checkNotNull(backoffPolicyProvider, "backoffPolicyProvider"));
+  }
+
+  @Override
+  void start() {
+    logger.log(Level.INFO, "Starting XdsClient {0}", XdsClientImpl.this);
+    adsLifecycleManager.start();
+  }
+
+  @Override
+  void shutdown() {
+    channel.shutdown();
+    adsLifecycleManager.shutdown();
+  }
+
+  @Override
+  void watchEndpointData(String cluster, EndpointWatcher endpointWatcher) {
+    endpointWatchers.addWatcher(cluster, endpointWatcher);
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("xdsNode", xdsNode).add("authority", authority).toString();
+  }
+
+  /**
+   * Keeps track of all the endpoinWaters that have been added.
+   */
+  static final class EndpointWatchers {
+    private final Map<String, Collection<EndpointWatcher>> endpointWatchers = new LinkedHashMap<>();
+    private final Node xdsNode;
+
+    @CheckForNull // It is null during backoff period.
+    private StreamObserver<DiscoveryRequest> xdsRequestWriter;
+
+    EndpointWatchers(Node xdsNode) {
+      this.xdsNode = xdsNode;
+    }
+
+    void addWatcher(String cluster, EndpointWatcher endpointWatcher) {
+      if (!endpointWatchers.containsKey(cluster)) {
+        endpointWatchers.put(cluster, new ArrayList<EndpointWatcher>());
+      }
+      endpointWatchers.get(cluster).add(endpointWatcher);
+
+      if (xdsRequestWriter != null) {
+        requestEndpoints(xdsRequestWriter);
+      }
+    }
+
+    void onError(Status error) {
+      for (Collection<EndpointWatcher> watcherCollection : endpointWatchers.values()) {
+        for (EndpointWatcher endpointWatcher : watcherCollection) {
+          endpointWatcher.onError(error);
+        }
+      }
+    }
+
+    void onEndpointUpdate(ClusterLoadAssignment clusterLoadAssignment) {
+      String cluster = clusterLoadAssignment.getClusterName();
+      if (endpointWatchers.containsKey(cluster)) {
+        for (EndpointWatcher endpointWatcher : endpointWatchers.get(cluster)) {
+          endpointWatcher.onEndpointChanged(
+              EndpointUpdate.newBuilder().setClusterLoadAssignment(clusterLoadAssignment).build());
+        }
+      }
+    }
+
+    /**
+     * Sends an EDS request if there is any EndpointWatcher added, otherwise will delay EDS request
+     * until a new EndpointWatcher is added.
+     */
+    void requestEndpoints(StreamObserver<DiscoveryRequest> xdsRequestWriter) {
```

Comment:
It will be good to have 2 separate methods as follows:

```
void setXdsRequestWriter(StreamObserver<DiscoveryRequest> xdsRequestWriter) {
      this.xdsRequestWriter = checkNotNull(xdsRequestWriter, "xdsRequestWriter");
      requestEndpoints();
}

void requestEndpoints() {
    if (requestWriter != null && !endpointWatchers.isEmpty()) {
        DiscoveryRequest edsRequest =
            DiscoveryRequest.newBuilder()
                .setNode(xdsNode)
                .setTypeUrl(EDS_TYPE_URL)
                .addAllResourceNames(endpointWatchers.keySet())
                .build();
        logger.log(Level.FINE, "Sending EDS request {0}", edsRequest);
        xdsRequestWriter.onNext(edsRequest);
      }
}
```

Then `AdsLifecycleManager.start()` can call `setXdsRequestWriter` and `XdsClientImpl.EndpointWatchers.addWatcher` can just call `requestEndpoints()`. Much cleaner...

Reply:
Then we also need get rid of `seizeRequestEndpoints()`. There might be two choices:
```java
void setXdsRequestWriter(@Nullable XdsRequestWriter writer);
```
and
```java
void setXdsRequestWriter(XdsRequestWriter writer);
void unsetXdsRequestWriter();
```
Which one do you favor? Or any other idea?

Reply:
I favor the `unset...` version. Sorry I didn't notice `seizeRequestEndpoints()` . Also should it have been `cease...` ?

Reply:
Yes, I had typos, and copied to everywhere.